openmc/cmfd.py

```diff
@@ -0,0 +1,3072 @@
+        if openmc.lib.master():
+            outstr = ("=====================>     "
+                      "CMFD TIMING STATISTICS     <====================\n\n"
+                      "   Time in CMFD                    =  {:.5E} seconds\n"
+                      "     Building matrices             =  {:.5E} seconds\n"
+                      "     Solving matrices              =  {:.5E} seconds\n")
+            print(outstr.format(self._time_cmfd, self._time_cmfdbuild,
+                                self._time_cmfdsolve))
+            sys.stdout.flush()
+
+    def _configure_cmfd(self):
+        """Initialize CMFD parameters and set CMFD input variables"""
+        # Check if restarting simulation from statepoint file
+        if not openmc.lib.settings.restart_run:
+            # Read in cmfd input defined in Python
+            self._read_cmfd_input()
+
+            # Set up CMFD coremap
+            self._set_coremap()
+
+            # Extract spatial and energy indices
+            nx, ny, nz, ng = self._indices
+
+            # Allocate parameters that need to stored for tally window
+            self._openmc_src_rate = np.zeros((nx, ny, nz, ng, 0))
+            self._flux_rate = np.zeros((nx, ny, nz, ng, 0))
+            self._total_rate = np.zeros((nx, ny, nz, ng, 0))
+            self._p1scatt_rate = np.zeros((nx, ny, nz, ng, 0))
+            self._scatt_rate = np.zeros((nx, ny, nz, ng, ng, 0))
+            self._nfiss_rate = np.zeros((nx, ny, nz, ng, ng, 0))
+            self._current_rate = np.zeros((nx, ny, nz, 12, ng, 0))
+
+            # Initialize timers
+            self._time_cmfd = 0.0
+            self._time_cmfdbuild = 0.0
+            self._time_cmfdsolve = 0.0
+
+            # Initialize parameters for CMFD tally windows
+            self._set_tally_window()
+
+        else:
+            # Reset CMFD parameters from statepoint file
+            path_statepoint = openmc.lib.settings.path_statepoint
+            self._reset_cmfd(path_statepoint)
+
+    def _read_cmfd_input(self):
+        """Sets values of additional instance variables based on user input"""
+        # Print message to user and flush output to stdout
+        if openmc.lib.settings.verbosity >= 7 and openmc.lib.master():
+            print(' Configuring CMFD parameters for simulation')
+            sys.stdout.flush()
+
+        # Check if CMFD mesh is defined
+        if self._mesh is None:
+            raise ValueError('No CMFD mesh has been specified for '
+                             'simulation')
+
+        # Set spatial dimensions of CMFD object
+        for i, n in enumerate(self._mesh.dimension):
+            self._indices[i] = n
+
+        # Check if in continuous energy mode
+        if not openmc.lib.settings.run_CE:
+            raise OpenMCError('CMFD must be run in continuous energy mode')
+
+        # Set number of energy groups
+        if self._mesh.energy is not None:
+            ng = len(self._mesh.energy)
```

Comment:
Super nit: Slightly unclear identifier in context.  If `ng` represents "numbered group", I would work energy in the variable name somewhere (maybe energy_group_count or nge?)